// @shelby-protocol/sdk/dist/browser/index.mjs
import "../chunk-DJJD2AXO.mjs";
import "../chunk-MWDW4ROU.mjs";
import "../chunk-MQUVYMNQ.mjs";
import "../chunk-ZHXCVRZX.mjs";
import "../chunk-RNXGC54D.mjs";
import {
  ShelbyClient
} from "../chunk-AUGZMI6U.mjs";
import "../chunk-Z7RFCADT.mjs";
import {
  ShelbyRPCClient
} from "../chunk-7P3P3D3X.mjs";
import {
  isShelbyNetwork,
  shelbyNetworks
} from "../chunk-VRLIOKWG.mjs";
import "../chunk-I6NG5GNL.mjs";
import {
  getShelbyAccountBlobsExplorerUrl,
  getShelbyAccountExplorerUrl,
  getShelbyBlobExplorerUrl
} from "../chunk-BDSW5PHM.mjs";
import {
  ShelbyBlobClient
} from "../chunk-HDYY6NS4.mjs";
import {
  createShelbyIndexerClient,
  getShelbyIndexerClient
} from "../chunk-HFGEQP5N.mjs";
import {
  Blob_Activities_Select_Column,
  Blobs_Select_Column,
  Cursor_Ordering,
  GetBlobActivitiesCountDocument,
  GetBlobActivitiesDocument,
  GetBlobsCountDocument,
  GetBlobsDocument,
  Order_By,
  Placement_Group_Slots_Select_Column,
  Processor_Status_Select_Column,
  getSdk
} from "../chunk-WJKSPJSS.mjs";
import "../chunk-RBFWGDMY.mjs";
import {
  getAptosAccountExplorerUrl,
  getAptosTransactionExplorerUrl
} from "../chunk-C6RQ3AEU.mjs";
import {
  createBlobKey
} from "../chunk-OTBLZL2S.mjs";
import {
  BlobCommitmentsSchema,
  ChunksetCommitmentSchema,
  expectedTotalChunksets,
  generateCommitments
} from "../chunk-PCNLFNAT.mjs";
import {
  buildRequestUrl,
  concatHashes,
  getBlobNameSuffix,
  readInChunks,
  zeroPadBytes
} from "../chunk-4JZO2D7T.mjs";
import {
  DEFAULT_PROJECT_DESCRIPTION,
  DEFAULT_PROJECT_NAME,
  NetworkToShelbyBlobIndexerBaseUrl,
  NetworkToShelbyRPCBaseUrl,
  SHELBYUSD_FA_METADATA_ADDRESS,
  SHELBYUSD_TOKEN_ADDRESS,
  SHELBYUSD_TOKEN_MODULE,
  SHELBYUSD_TOKEN_NAME,
  SHELBY_DEPLOYER,
  TOKEN_DEPLOYER,
  TOKEN_OBJECT_ADDRESS
} from "../chunk-SEXQTDX6.mjs";
import {
  BlobNameSchema,
  ChunkKey,
  allChunksForBlob,
  roundSize
} from "../chunk-FIFKKWXV.mjs";
import {
  CHUNK_SIZE_PARAMS,
  ChunkSizeScheme,
  ClayErasureCodingProvider,
  DEFAULT_CHUNKSET_SIZE_BYTES,
  DEFAULT_CHUNK_SIZE_BYTES,
  ERASURE_CODE_AND_CHUNK_MAPPING,
  createDefaultErasureCodingProvider
} from "../chunk-APML3CGJ.mjs";
import {
  ReedSolomonErasureCodingProvider
} from "../chunk-3ZDXWPYC.mjs";
import {
  DEFAULT_ERASURE_D,
  DEFAULT_ERASURE_K,
  DEFAULT_ERASURE_M,
  DEFAULT_ERASURE_N,
  ERASURE_CODE_PARAMS,
  ErasureCodingScheme
} from "../chunk-ZPW742E7.mjs";
import "../chunk-7P6ASYW6.mjs";
export {
  BlobCommitmentsSchema,
  BlobNameSchema,
  Blob_Activities_Select_Column,
  Blobs_Select_Column,
  CHUNK_SIZE_PARAMS,
  ChunkKey,
  ChunkSizeScheme,
  ChunksetCommitmentSchema,
  ClayErasureCodingProvider,
  Cursor_Ordering,
  DEFAULT_CHUNKSET_SIZE_BYTES,
  DEFAULT_CHUNK_SIZE_BYTES,
  DEFAULT_ERASURE_D,
  DEFAULT_ERASURE_K,
  DEFAULT_ERASURE_M,
  DEFAULT_ERASURE_N,
  DEFAULT_PROJECT_DESCRIPTION,
  DEFAULT_PROJECT_NAME,
  ERASURE_CODE_AND_CHUNK_MAPPING,
  ERASURE_CODE_PARAMS,
  ErasureCodingScheme,
  GetBlobActivitiesCountDocument,
  GetBlobActivitiesDocument,
  GetBlobsCountDocument,
  GetBlobsDocument,
  NetworkToShelbyBlobIndexerBaseUrl,
  NetworkToShelbyRPCBaseUrl,
  Order_By,
  Placement_Group_Slots_Select_Column,
  Processor_Status_Select_Column,
  ReedSolomonErasureCodingProvider,
  SHELBYUSD_FA_METADATA_ADDRESS,
  SHELBYUSD_TOKEN_ADDRESS,
  SHELBYUSD_TOKEN_MODULE,
  SHELBYUSD_TOKEN_NAME,
  SHELBY_DEPLOYER,
  ShelbyBlobClient,
  ShelbyClient,
  ShelbyRPCClient,
  TOKEN_DEPLOYER,
  TOKEN_OBJECT_ADDRESS,
  allChunksForBlob,
  buildRequestUrl,
  concatHashes,
  createBlobKey,
  createDefaultErasureCodingProvider,
  createShelbyIndexerClient,
  expectedTotalChunksets,
  generateCommitments,
  getAptosAccountExplorerUrl,
  getAptosTransactionExplorerUrl,
  getBlobNameSuffix,
  getSdk,
  getShelbyAccountBlobsExplorerUrl,
  getShelbyAccountExplorerUrl,
  getShelbyBlobExplorerUrl,
  getShelbyIndexerClient,
  isShelbyNetwork,
  readInChunks,
  roundSize,
  shelbyNetworks,
  zeroPadBytes
};
